src/librbd/ManagedLock.h

```diff
@@ -0,0 +1,181 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MANAGED_LOCK_H
+#define CEPH_LIBRBD_MANAGED_LOCK_H
+
+#include "include/int_types.h"
+#include "include/Context.h"
+#include "include/rados/librados.hpp"
+#include "cls/lock/cls_lock_types.h"
+#include "common/Mutex.h"
+#include <list>
+#include <string>
+#include <utility>
+
+class ContextWQ;
+
+namespace librbd {
+
+class Watcher;
+
+template <typename ImageCtxT>
+class ManagedLock {
+public:
+  static const std::string WATCHER_LOCK_TAG;
+
+  static ManagedLock *create(librados::IoCtx& ioctx, ContextWQ *work_queue,
+                             const std::string& oid, Watcher *watcher) {
+    return new ManagedLock(ioctx, work_queue, oid, watcher);
+  }
+
+  ManagedLock(librados::IoCtx& ioctx, ContextWQ *work_queue,
+              const std::string& oid, Watcher *watcher);
+  virtual ~ManagedLock();
+
+  bool is_lock_owner() const;
+
+  void shut_down(Context *on_shutdown);
+  void acquire_lock(Context *on_acquired, bool try_lock = false);
```

Comment:
Minor: really just my personal preference, but I really dislike defaulted variables. I think it would be clearer to have a ```try_acquire_lock``` method and a ```acquire_lock``` method since you need to differentiate the states.